haskell/cabal.bzl

```diff
@@ -892,23 +892,71 @@ def _get_components(components, package):
     """
     return components.get(package, _default_components.get(package, struct(lib = True, exe = [])))
 
-def _validate_package_specs(package_specs):
-    found_ty = type(package_specs)
-    if found_ty != "list":
-        fail("Unexpected output format for `stack ls dependencies json`. Expected 'list', but got '%s'." % found_ty)
-
-def _validate_package_spec(package_spec):
-    fields = [
-        ("name", "string"),
-        ("version", "string"),
-        ("dependencies", "list"),
-    ]
-    for (field, ty) in fields:
-        if not field in package_spec:
-            fail("Unexpected output format for `stack ls dependencies json`. Missing field '%s'." % field)
-        found_ty = type(package_spec[field])
-        if found_ty != ty:
-            fail("Unexpected output format for `stack ls dependencies json`. Expected field '%s' of type '%s', but got type '%s'." % (field, ty, found_ty))
+def _parse_json_field(json, field, ty, fmt):
+    """Read and type-check a field from a JSON object.
+
+    Invokes `fail` on error.
+
+    Attrs:
+      json: dict, The parsed JSON object.
+      field: string, The name of the field.
+      ty: string, The expected type of the field.
+      fmt: string, Error message format string. E.g. `Error: %s`.
+
+    Returns:
+      The value of the field.
+    """
+    if not field in json:
+        fail(fmt % "Missing field '{field}'.".format(field = field))
+    actual_ty = type(json[field])
+    if actual_ty != ty:
+        fail(fmt % "Expected field '{field}' of type '{expected}', but got '{got}'.".format(
+            field = field,
+            expected = ty,
+            got = actual_ty,
+        ))
+    return json[field]
+
+def _parse_package_spec(package_spec):
```

Comment:
A link to the definition of the package spec format would be nice (though I don’t have the illusion that there exists one)

Reply:
That's a good idea. I'm not aware of a standalone definition, however the `stack` sources certainly define the current format. I've added a link to the docstring.